### DISCUSSION TITLE
Crazyflie landing code

### DISCUSSION THREAD

**Discussion by crazy-interns** [bitcraze/discussions#1585] [Q&A]
I have written a code to just simply make the drone go up, hover and land. It does go up just fine but it does not land smoothly. the drone motors just stops midair and it abruptly falls down everytime i have made them fly using any code. PLease how can I modify the code so the motors stop working after the drone has landed. 

My code:

```python
# -*- coding: utf-8 -*-  # Specify the encoding of the file

# Import necessary libraries
import time  # For time-related functions (e.g., sleep)
import cflib.crtp  # Import the Crazyflie driver library for communication
from cflib.crazyflie import Crazyflie  # Import the Crazyflie class for handling Crazyflie drone
from cflib.crazyflie.high_level_commander import HighLevelCommander  # Import commander for high-level commands
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie  # For synchronous communication with Crazyflie
from cflib.utils import uri_helper  # Helper functions for URI handling


# Define the URI to connect to the Crazyflie drone
uri = uri_helper.uri_from_env(default='radio://0/100/2M/E7E7E7E7E7')
# This sets a default URI if not specified in the environment; it connects via radio with specific settings.

def run_sequence(cf):
    # This function will run a sequence of commands for the Crazyflie drone
    commander = cf.high_level_commander  # Get the high-level commander object to send commands

    # Command to take off to 20 cm (0.2 meters) over 2 seconds
    commander.takeoff(1.0, 10.0)  # the second parameter means: time taken for drone to takeoff/land
    time.sleep(5)  # Pause execution for 5 seconds to allow the drone to hover
    # Command to land the drone

    commander.land(0.0, 10.0)  # Land at height of 0.0, duration in seconds
    time.sleep(2)  # Pause for 2 seconds to let the landing stabilize
    commander.stop()  # Stop any ongoing commands to ensure the drone is not executing further movements


if __name__ == '__main__':
    # Check if this script is run as the main program
    cflib.crtp.init_drivers()  # Initialize the Crazyflie drivers needed for communication

    # Create a synchronous Crazyflie object to manage the connection
    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:
        # Using the URI, connect to the Crazyflie and maintain a session
        cf = scf.cf  # Get the Crazyflie instance from the sync context
        run_sequence(cf)  # Call the function to execute the takeoff, hover, and landing sequence
```

**Comment by gemenerik** [comment]
There's a discrepancy between the time you sleep the Python program and the duration of your commands. The Python program does not block while the commander commands are executing, meaning that `commander.stop` will execute 7 seconds after your takeoff command.

**Reply by crazy-interns** [reply]
I wanted to ask you. Do you know why does my drone just flip over every time I run a code with it? like I have 5 of them and I find really difficult to fly them even though I have lighthouse deck and its all set and everything. The drones were more cooperative before but now they just flip over, or 2 of the motors just stop rotating midway. Is this a common issue? Should I change my propellers to better ones? because they have been breaking too as it suddenly hits the wall or ground. the drones are very unpredictable. There doesnt seem to be any hardware issue. like if a run a code twice both times the drones react differently even though it is the same code. should i change the motors?

**Reply by gemenerik** [reply]
I'm sorry to hear you're having these issues, and I understand how frustrating it is, especially with propellers breaking from unexpected flips and crashes. Are you noticing this behavior with all your drones, or just a few? Since you mentioned they used to be more cooperative, I'm curious if there have been any changes to the software, hardware, or even the environment that might be affecting them.

One key thing to check is whether the Lighthouse system is properly calibrated. Also, ensure there are no reflective surfaces nearby, as these can interfere with the Lighthouse signals and cause unpredictable behavior.

**Reply by gemenerik** [reply]
Another culprit could be communication. How did you set up your URIs and how many radios are you using? Do you see the same problems when flying a single drone? Are you using the commander from the cflib like in the above example?

**Comment by crazy-interns** [comment]
We didn't write anything onto the crazyflies firmware. We use 1 radio and
we did change the uri for each drone so we can differentiate them. Even
when flying a single drone I faced this problem. Even when I connect a
single drone to cfclient and fly it through their take off button, the
drone just doesn't fly smoothly. Before they used to sometimes but now all
of them are just either flipping or don't fly. I updated each of their
firmware to latest, the hardware in all seems good. The lighthouse seems to
work just fine. Yet the drones don't fly properly whether it's with code or
cfclient. One of them was flipping then I updated its firmware in cold boot
then tried doing take off from cfclient it flew and hovered smoothly. Then
again after I restarted it, it started flipping over again.

**Reply by gemenerik** [reply]
Even if the Lighthouse system seems operational, please recalibrate it, especially if there were recent changes to the environment. Even subtle changes can interfere with the localization, leading to unexpected flips or instability. Make sure there are no reflective surfaces at all. I've had problems with something as small as a metal chair leg. To check the positioning quality without damaging your drones, you can review the Lighthouse client tab while walking around with the drone in your hand. Make sure the positioning is stable throughout your flight area.

Weak or unstable battery power can cause motors to lose power mid-flight. If you haven't already, try replacing the batteries.

When using a single Crazyradio to control multiple Crazyflies, make sure to keep them on the same channel and different addresses. 

Can you fly it through other means? Like using the take-off and direction buttons in the client?

**Comment by crazy-interns** [comment]
I only have 1 base station 2. I was able to calibrate using 1 base station
v2 but I can't find the link where I saw the steps from so I can
recalibrate. Do you know where I can find it?.

Also for the take off using client yes I was able to fly them before but
now they just flip. I think I will try without any reflective surface as
well.

**Reply by Bananazooka** [reply]
Hi, you can find the steps to get a new geometry estimation [here](https://www.bitcraze.io/documentation/tutorials/getting-started-with-lighthouse/#preparing-the-system).

**Reply by crazy-interns** [reply]
this is for 2 base stations. i only have 1 base station and i am not able to click on estimate geometry either of the buttons

**Reply by knmcguire** [reply]
Hi! 

Yes that is a special case. you can use the geometry estimation simple for that. The instructions for that can be found in the tab documentation here: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/lighthouse_tab/#basestation-geometry-management

**Comment by cuber-1** [comment]
Hi, I'm trying to code my first script without using any decks but the AI Deck. I've been trying to find a way to control the robot with code that doesn't require a flow deck (or something similar). How exactly would you run that script you've written? Thanks!

**Reply by sakina-k-ali** [reply]
Hello, I would suggest you try out droneblocks app to fly your drone without any decks using their inbuilt code but if you are to code personally and then fly them then a positioning system would be necessary in my opinion to keep the drone stable and smooth.

**Reply by cuber-1** [reply]
If they have inbuilt code to be able to fly the drone without sensors there should be a way for me to write that as well, right?

**Reply by ToveRumar** [reply]
Hi @cuber-1 !

Flying using an app or gamepad without any decks is definitely possible. However this is only possible since you are able to observe the drone and in real time and compensate for any drifting other movements. You are in this case acting as its sensors. 

This can not be done using a script. A script can not compensate any drifting or unexpected movements since this would be different everytime you fly it. 

If you are thinking about how other drones can fly stably, this is due to them having many on board sensors. For the Crazyflie you can customize what sensors you want, by equipping it with decks. 

I hope this clarified why this is not possible to do without any additional sensors.

**Reply by sakina-k-ali** [reply]
Hi, so I really need to do a simple show with my crazyflie shows before monday as a project and the drones were working completely fine and now anything i do, the drones don't fly. they go up a bit and then flip. I have been using the lighthouse positioning system and have 5 drones. only 1 of them works with cfclient take off and land + any python code to make it fly but rest just flip even though their LPS decks are all configured correctly. since 1 of the drone functions correctly, i know the issue is not because of the LPS but rather the drones themselves. I connected and ran their console tab in cfclient and something called "assertfail" or normalstart [FAIL] shows up. I don't remember writing anything onto the drones deck or anything to corrupt them. the decks are green and works fine but the drones themselves keep flipping. please is there any code to just make the drones go back to their original settings since updating firmware, coldbooting nothing has worked for me at all. Please i need an urgent reply @ToveRumar @gemenerik
